src/coprocessor/dag/rpn_expr/types/function.rs

```diff
@@ -79,35 +71,51 @@
 //! If you are curious about what code the macro will generate, check the test code
 //! in `components/cop_codegen/src/rpn_function.rs`.
 
-use super::types::{RpnFnCallPayload, RpnStackNode};
+use tipb::expression::FieldType;
+
+use super::RpnStackNode;
 use crate::coprocessor::codec::data_type::{Evaluable, ScalarValue, VectorValue};
 use crate::coprocessor::dag::expr::EvalContext;
 use crate::coprocessor::Result;
 
+/// Metadata of an RPN function.
 #[derive(Clone, Copy)]
-/// Metadata of an RPN function
 pub struct RpnFnMeta {
-    /// The display name of the function.
+    /// The display name of the RPN function. Mainly used in tests.
     pub name: &'static str,
 
-    /// The accepted argument length of this RPN function.
-    ///
-    /// Currently we do not support variable arguments.
-    pub args_len: usize,
+    /// Validator against input expression tree.
+    // pub validator_ptr: fn(expr: &Expr) -> Result<()>,
 
-    /// The function receiving raw argument.
-    ///
-    /// The first parameter is the evaluation context and the second one is the payload containing
-    /// the output rows count, the argument value and the argument field type.
-    pub fn_ptr: fn(&mut EvalContext, &RpnFnCallPayload<'_>) -> Result<VectorValue>,
+    #[allow(clippy::type_complexity)]
+    /// The RPN function.
+    pub fn_ptr: fn(
+        // Common arguments
+        ctx: &mut EvalContext,
+        output_rows: usize,
+        args: &[RpnStackNode<'_>],
+        // Uncommon arguments are grouped together
+        extra: &mut RpnFnCallExtra<'_>,
+        // Buffer for evaluating vargs.
+        vargs_buffer: &mut [usize],
```

Comment:
Do you think it a good idea to use a TLS buffer instead of passing it as an explicit parameter?

Reply:
I wish I could remove this explicit parameter.